File: Source/ARTTokenParams.m
//
//  ARTTokenParams.h
//  ably-ios
//
//  Created by Ricardo Pereira on 05/10/2015.
//  Copyright (c) 2015 Ably. All rights reserved.
//

#import "ARTTokenParams+Private.h"

#include <CommonCrypto/CommonDigest.h>
#include <CommonCrypto/CommonHMAC.h>

#import "ARTDefault.h"
#import "ARTEncoder.h"
#import "ARTTokenRequest.h"

@implementation ARTTokenParams

- (instancetype)init {
    return [self initWithClientId:nil nonce:nil];
}

- (instancetype)initWithClientId:(NSString *)clientId {
    return [self initWithClientId:clientId nonce:nil];
}

- (instancetype)initWithClientId:(NSString *)clientId nonce:(NSString *)nonce {
    if (self = [super init]) {
        _timestamp = nil;
        _capability = nil; // capabilities of the underlying key
        _clientId = clientId;
        _nonce = nonce;
    }
    return self;
}

- (instancetype)initWithOptions:(ARTClientOptions *)options {
    self = [self initWithClientId:options.clientId];
    if (options.defaultTokenParams) {
        if (options.defaultTokenParams.ttl) _ttl = options.defaultTokenParams.ttl;
        if (options.defaultTokenParams.capability) _capability = options.defaultTokenParams.capability;
    }
    return self;
}

- (instancetype)initWithTokenParams:(ARTTokenParams *)tokenParams {
    self = [self initWithClientId:tokenParams.clientId];
    self.timestamp = nil;
    self.ttl = tokenParams.ttl;
    self.capability = tokenParams.capability;
    return self;
}

- (NSString *)description {
    return [NSString stringWithFormat: @"ARTTokenParams: ttl=%@ capability=%@ timestamp=%@",
            self.ttl, self.capability, self.timestamp];
}

- (id)copyWithZone:(NSZone *)zone {
    ARTTokenParams *token = [[[self class] allocWithZone:zone] initWithClientId:_clientId nonce:_nonce];

    token.ttl = _ttl;
    token.capability = _capability;
    token.timestamp = _timestamp;

    return token;
}

- (NSMutableArray *)toArray {
    NSMutableArray *params = [[NSMutableArray alloc] init];
    
    if (self.clientId)
        [params addObject:[NSURLQueryItem queryItemWithName:@"clientId" value:self.clientId]];
    if (self.ttl)
        [params addObject:[NSURLQueryItem queryItemWithName:@"ttl" value:[NSString stringWithFormat:@"%@", self.ttl]]];
    if (self.capability)
        [params addObject:[NSURLQueryItem queryItemWithName:@"capability" value:self.capability]];
    if (self.timestamp)
        [params addObject:[NSURLQueryItem queryItemWithName:@"timestamp" value:[NSString stringWithFormat:@"%llu", dateToMilliseconds(self.timestamp)]]];
    
    return params;
}

- (NSMutableDictionary *)toDictionary {
    NSMutableDictionary *params = [[NSMutableDictionary alloc] init];
    
    if (self.clientId)
        params[@"clientId"] = self.clientId;
    if (self.ttl)
        params[@"ttl"] = [NSString stringWithFormat:@"%@", self.ttl];
    if (self.capability)
        params[@"capability"] = self.capability;
    if (self.timestamp)
        params[@"timestamp"] = [NSString stringWithFormat:@"%llu", dateToMilliseconds(self.timestamp)];
    
    return params;
}

- (NSArray *)toArrayWithUnion:(NSArray *)items {
    NSMutableArray *tokenParams = [self toArray];
    BOOL add = YES;
    
    for (NSURLQueryItem *item in items) {
        for (NSURLQueryItem *param in tokenParams) {
            // Check if exist
            if ([param.name isEqualToString:item.name]) {
                add = NO;
                break;
            }
        }
        if (add) {
            [tokenParams addObject:item];
        }
        add = YES;
    }
    
    return tokenParams;
}

- (NSDictionary *)toDictionaryWithUnion:(NSArray *)items {
    NSMutableDictionary *tokenParams = [self toDictionary];
    BOOL add = YES;
    
    for (NSURLQueryItem *item in items) {
        for (NSString *key in tokenParams.allKeys) {
            // Check if exist
            if ([key isEqualToString:item.name]) {
                add = NO;
                break;
            }
        }
        if (add) {
            tokenParams[item.name] = item.value;
        }
        add = YES;
    }
    
    return tokenParams;
}

static NSString *hmacForDataAndKey(NSData *data, NSData *key) {
    const void *cKey = [key bytes];
    const void *cData = [data bytes];
    size_t keyLen = [key length];
    size_t dataLen = [data length];
    
    unsigned char hmac[CC_SHA256_DIGEST_LENGTH];
    
    CCHmac(kCCHmacAlgSHA256, cKey, keyLen, cData, dataLen, hmac);
    NSData *mac = [[NSData alloc] initWithBytes:hmac length:sizeof(hmac)];
    NSString *str = [mac base64EncodedStringWithOptions:0];
    return str;
}

- (ARTTokenRequest *)sign:(NSString *)key {
    return [self sign:key withNonce:self.nonce ? self.nonce : generateNonce()];
}

- (ARTTokenRequest *)sign:(NSString *)key withNonce:(NSString *)nonce {
    NSArray *keyComponents = decomposeKey(key);
    NSString *keyName = keyComponents[0];
    NSString *keySecret = keyComponents[1];
    NSString *capability = self.capability ? self.capability : @"";
    NSString *clientId = self.clientId ? self.clientId : @"";
    NSString *ttl = self.ttl ? [NSString stringWithFormat:@"%lld", timeIntervalToMilliseconds([self.ttl doubleValue])] : @"";

    NSString *signText = [NSString stringWithFormat:@"%@\n%@\n%@\n%@\n%lld\n%@\n", keyName, ttl, capability, clientId, dateToMilliseconds(self.timestamp), nonce];
    NSString *mac = hmacForDataAndKey([signText dataUsingEncoding:NSUTF8StringEncoding], [keySecret dataUsingEncoding:NSUTF8StringEncoding]);

    return [[ARTTokenRequest alloc] initWithTokenParams:self keyName:keyName nonce:nonce mac:mac];
}

@end

File: Sources/AblySwift/ARTTokenParams.swift
import Foundation
import CommonCrypto

// swift-migration: original location ARTTokenParams.h, line 13 and ARTTokenParams.m, line 10
/// Defines the properties of an Ably Token.
public class ARTTokenParams: NSObject, NSCopying {
    
    // swift-migration: original location ARTTokenParams.h, line 18
    /// Requested time to live for the token in milliseconds. The default is 60 minutes.
    public var ttl: NSNumber?
    
    // swift-migration: original location ARTTokenParams.h, line 23
    /// The capabilities associated with this Ably Token. The capabilities value is a JSON-encoded representation of the resource paths and associated operations. Read more about capabilities in the [capabilities docs](https://ably.com/docs/core-features/authentication/#capabilities-explained).
    public var capability: String?
    
    // swift-migration: original location ARTTokenParams.h, line 28
    /// A client ID, used for identifying this client when publishing messages or for presence purposes. The `clientId` can be any non-empty string, except it cannot contain a `*`. This option is primarily intended to be used in situations where the library is instantiated with a key. Note that a `clientId` may also be implicit in a token used to instantiate the library. An error is raised if a `clientId` specified here conflicts with the `clientId` implicit in the token. Find out more about [identified clients](https://ably.com/docs/core-features/authentication#identified-clients).
    public var clientId: String?
    
    // swift-migration: original location ARTTokenParams.h, line 33
    /// The timestamp of this request as `NSDate` object. Timestamps, in conjunction with the `nonce`, are used to prevent requests from being replayed. `timestamp` is a "one-time" value, and is valid in a request, but is not validly a member of any default token params such as `ARTClientOptions.defaultTokenParams`.
    public var timestamp: Date?
    
    // swift-migration: original location ARTTokenParams.h, line 38
    /// A cryptographically secure random string of at least 16 characters, used to ensure the `ARTTokenRequest` cannot be reused.
    public var nonce: String?
    
    // swift-migration: original location ARTTokenParams.m, line 12
    public required override init() {
        super.init()
        self.timestamp = nil
        self.capability = nil
        self.clientId = nil
        self.nonce = nil
    }
    
    // swift-migration: original location ARTTokenParams.h, line 44 and ARTTokenParams.m, line 16
    public convenience init(clientId: String?) {
        self.init()
        self.clientId = clientId
    }
    
    // swift-migration: original location ARTTokenParams.h, line 47 and ARTTokenParams.m, line 20
    public convenience init(clientId: String?, nonce: String?) {
        self.init()
        self.clientId = clientId
        self.nonce = nonce
    }
    
    // swift-migration: original location ARTTokenParams.h, line 50 and ARTTokenParams.m, line 30
    public init(options: ARTClientOptions) {
        super.init()
        self.timestamp = nil
        self.capability = nil
        self.clientId = options.clientId
        self.nonce = nil
        
        if let defaultTokenParams = options.defaultTokenParams {
            if let ttl = defaultTokenParams.ttl {
                self.ttl = ttl
            }
            if let capability = defaultTokenParams.capability {
                self.capability = capability
            }
        }
    }
    
    // swift-migration: original location ARTTokenParams.h, line 53 and ARTTokenParams.m, line 39
    public init(tokenParams: ARTTokenParams) {
        super.init()
        self.clientId = tokenParams.clientId
        self.timestamp = nil
        self.ttl = tokenParams.ttl
        self.capability = tokenParams.capability
        self.nonce = tokenParams.nonce
    }
    
    // swift-migration: original location ARTTokenParams.m, line 47
    public override var description: String {
        return "ARTTokenParams: ttl=\(String(describing: ttl)) capability=\(String(describing: capability)) timestamp=\(String(describing: timestamp))"
    }
    
    // swift-migration: original location ARTTokenParams.m, line 52
    public func copy(with zone: NSZone?) -> Any {
        let token = type(of: self).init()
        token.clientId = clientId
        token.nonce = nonce
        token.ttl = ttl
        token.capability = capability
        token.timestamp = timestamp
        return token
    }
    
    // swift-migration: original location ARTTokenParams.h, line 56 and ARTTokenParams.m, line 62
    public func toArray() -> [URLQueryItem] {
        var params: [URLQueryItem] = []
        
        if let clientId = self.clientId {
            params.append(URLQueryItem(name: "clientId", value: clientId))
        }
        if let ttl = self.ttl {
            params.append(URLQueryItem(name: "ttl", value: "\(ttl)"))
        }
        if let capability = self.capability {
            params.append(URLQueryItem(name: "capability", value: capability))
        }
        if let timestamp = self.timestamp {
            params.append(URLQueryItem(name: "timestamp", value: "\(dateToMilliseconds(timestamp))"))
        }
        
        return params
    }
    
    // swift-migration: original location ARTTokenParams.m, line 77
    internal func toDictionary() -> [String: String] {
        var params: [String: String] = [:]
        
        if let clientId = self.clientId {
            params["clientId"] = clientId
        }
        if let ttl = self.ttl {
            params["ttl"] = "\(ttl)"
        }
        if let capability = self.capability {
            params["capability"] = capability
        }
        if let timestamp = self.timestamp {
            params["timestamp"] = "\(dateToMilliseconds(timestamp))"
        }
        
        return params
    }
    
    // swift-migration: original location ARTTokenParams.h, line 59 and ARTTokenParams.m, line 92
    public func toArray(withUnion items: [URLQueryItem]) -> [URLQueryItem] {
        var tokenParams = toArray()
        var add = true
        
        for item in items {
            for param in tokenParams {
                // Check if exist
                if param.name == item.name {
                    add = false
                    break
                }
            }
            if add {
                tokenParams.append(item)
            }
            add = true
        }
        
        return tokenParams
    }
    
    // swift-migration: original location ARTTokenParams.h, line 62 and ARTTokenParams.m, line 113
    public func toDictionary(withUnion items: [URLQueryItem]) -> [String: String] {
        var tokenParams = toDictionary()
        var add = true
        
        for item in items {
            for key in tokenParams.keys {
                // Check if exist
                if key == item.name {
                    add = false
                    break
                }
            }
            if add {
                if let value = item.value {
                    tokenParams[item.name] = value
                }
            }
            add = true
        }
        
        return tokenParams // immutable copy in Swift
    }
}

// swift-migration: original location ARTTokenParams.m, line 134
internal func hmacForDataAndKey(_ data: Data, _ key: Data) -> String {
    let cKey = key.withUnsafeBytes { $0.bindMemory(to: UInt8.self).baseAddress! }
    let cData = data.withUnsafeBytes { $0.bindMemory(to: UInt8.self).baseAddress! }
    let keyLen = key.count
    let dataLen = data.count
    
    var hmac = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
    
    CCHmac(CCHmacAlgorithm(kCCHmacAlgSHA256), cKey, keyLen, cData, dataLen, &hmac)
    let mac = Data(bytes: hmac, count: hmac.count)
    return mac.base64EncodedString()
}

// MARK: - Private methods extension

extension ARTTokenParams {
    
    // swift-migration: original location ARTTokenParams+Private.h, line 5 and ARTTokenParams.m, line 148
    internal func sign(_ key: String) -> ARTTokenRequest {
        return sign(key, withNonce: nonce ?? generateNonce())
    }
    
    // swift-migration: original location ARTTokenParams+Private.h, line 6 and ARTTokenParams.m, line 152
    internal func sign(_ key: String, withNonce nonce: String) -> ARTTokenRequest {
        let keyComponents = decomposeKey(key)
        let keyName = keyComponents[0]
        let keySecret = keyComponents[1]
        let capability = self.capability ?? ""
        let clientId = self.clientId ?? ""
        let ttl = self.ttl != nil ? "\(timeIntervalToMilliseconds(self.ttl!.doubleValue))" : ""
        
        let signText = "\(keyName)\n\(ttl)\n\(capability)\n\(clientId)\n\(dateToMilliseconds(self.timestamp!))\n\(nonce)\n"
        let mac = hmacForDataAndKey(signText.data(using: .utf8)!, keySecret.data(using: .utf8)!)
        
        return ARTTokenRequest(tokenParams: self, keyName: keyName, nonce: nonce, mac: mac)
    }
}